### PR TITLE
New user input for grid type (collocated, staggered, hybrid)

### PR DESCRIPTION
**Pull request by EZoni**
- Old input config for hybrid mode (full hybrid, w/ fields and currents centering):
```sh
warpx.do_nodal = 0 #optional
algo.field_gathering = momentum-conserving #necessary, otherwise set to energy-conserving by default
interpolation.field_centering_nox = 8 #necessary, otherwise set to 2 by default
interpolation.field_centering_noy = 8 #necessary, otherwise set to 2 by default
interpolation.field_centering_noz = 8 #necessary, otherwise set to 2 by default
warpx.do_current_centering = 1 #necessary, otherwise set to 0 (false) by default
interpolation.current_centering_nox = 8 #necessary, otherwise set to 2 by default
interpolation.current_centering_noy = 8 #necessary, otherwise set to 2 by default
interpolation.current_centering_noz = 8 #necessary, otherwise set to 2 by default
```
- New input config for hybrid mode (full hybrid, w/ fields and currents centering):
```sh
warpx.grid_type = hybrid #necessary, otherwise set to staggered by default
```
- The new input parameter `warpx.grid_type` can be set to `collocated` (which corresponds to nodal currently, but could be changed to fully cell-centered in the future, e.g. to avoid symmetry breaking), `staggered` (which corresponds to purely staggered), and `hybrid`.
- The old input parameter `warpx.do_nodal` (0/1) was removed.
- If the user does want to tweak details of the hybrid algo, for instance switch off currents centering or change the order of the finite centering of fields and/or currents, all relevant parameters are now grouped under the same namespace `warpx`, for convenience: `warpx.field_centering_no<x,y,z>`, `warpx.current_centering_no<x,y,z>`, `warpx.do_current_centering`.
- The field gathering algo must be momentum-conserving, so fields centering is mandatory for the hybrid mode. On the other hand, currents centering is still optional.

To-do:
- [x] Fix bug: `grid_type` probably still used as `bool` somewhere
  (hybrid tests pass if `GridType::Hybrid` equals `0` instead of `2`)
- [x] Set default algo parameters for hybrid case
- [x] More changes to interface (e.g., all hybrid input parameters under `warpx` namespace)
- [x] Update examples with new user input syntax
- [x] Update documentation
- [x] Improve PR description, add details

### CODE REVIEW

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp [hunk 1]
```diff
@@ -70,11 +70,11 @@ void FiniteDifferenceSolver::EvolveB (
         ignore_unused(Gfield, face_areas);
         EvolveBCylindrical <CylindricalYeeAlgorithm> ( Bfield, Efield, lev, dt );
 #else
-    if(m_do_nodal or m_fdtd_algo != ElectromagneticSolverAlgo::ECT){
+    if(m_grid_type == GridType::Collocated or m_fdtd_algo != ElectromagneticSolverAlgo::ECT){
```
Comment (EZoni) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]:
```suggestion
    if(m_grid_type == GridType::Collocated || m_fdtd_algo != ElectromagneticSolverAlgo::ECT){
```

File: Source/WarpX.cpp [hunk 2]
```diff
@@ -1113,16 +1117,16 @@ WarpX::ReadParameters ()
         }
 
         // Finite-order centering is not implemented with mesh refinement
-        // (note that when WarpX::do_nodal = 1 finite-order centering is not used anyways)
-        if (maxLevel() > 0 && WarpX::do_nodal == 0)
+        // (note that when warpx.grid_type=collocated, finite-order centering is not used anyways)
+        if (maxLevel() > 0 && WarpX::grid_type == GridType::Hybrid)
```
Comment (EZoni) [inline Source/WarpX.cpp]:
```suggestion
        if (maxLevel() > 0 && WarpX::grid_type != GridType::Collocated)
```

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveE.cpp [hunk 3]
```diff
@@ -51,7 +51,7 @@ void FiniteDifferenceSolver::MacroscopicEvolveE (
         "currently macro E-push does not work for RZ"));
 #else
     WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
-        !m_do_nodal, "macro E-push does not work for nodal");
+        m_grid_type != GridType::Collocated, "macro E-push does not work for nodal");
```
Comment (EZoni) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveE.cpp]:
```suggestion
        m_grid_type != GridType::Collocated, "Macroscopic E field solver does not work on collocated grids");
```

File: Source/Particles/MultiParticleContainer.cpp [hunk 4]
```diff
@@ -1311,9 +1311,9 @@ MultiParticleContainer::doQEDSchwinger ()
 
     auto & warpx = WarpX::GetInstance();
 
-    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(warpx.do_nodal ||
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(warpx.grid_type ||
```
Comment (EZoni) [inline Source/Particles/MultiParticleContainer.cpp]:
```suggestion
    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(warpx.grid_type == GridType::Collocated ||
```

File: Source/Utils/WarpXAlgorithmSelection.H [hunk 5]
```diff
@@ -36,6 +36,15 @@
     };
 };
 
+struct GridType {
+    enum {
+        // FIXME
```
[the remark this answers is not in the document]
Comment (EZoni) [reply Source/Utils/WarpXAlgorithmSelection.H]: Fixed in 07d73e2.

**Comment by EZoni** [comment]
@RemiLehe @oshapoval @RTSandberg @n01r 
This PR should make it much easier to set up the hybrid mode correctly and automatically.
Please feel free to add your review and/or discuss other solutions either here or via Slack.

**Review by RTSandberg — APPROVED** [review]
Thanks @EZoni ! This will be useful to users like me who aren't expert in the various algorithmic and numeric options.  The naming is more descriptive and simpler to use.

**Comment by EZoni** [comment]
As pointed out by @RTSandberg, the PICMI interfaces should be updated as well, either here or in a follow-up PR.

It would require to change the interface of the initialization of the class `PICMI_ElectromagneticSolver`,
https://github.com/picmi-standard/picmi/blob/a010eee26b4fde95ac0db5254bd89d1c13c07346/PICMI_Python/fields.py#L76-L79,
by replacing `l_nodal` with something similar to WarpX's new `grid_type` parameter.

Discussed offline: the changes on PICMI will be done in a separate PR, we will keep the `l_nodal` flag for the electromagnetic solver and add a WarpX-specific input parameter for the hybrid mode. In other words, we will mimic the old hybrid user interface in WarpX, to keep the current PICMI standard interface as it is (since the hybrid algo is not "standard" yet).

**Review by RemiLehe — APPROVED** [review]

**Pull request merged**